src/server/zset_family.cc

```diff
@@ -1351,6 +1351,74 @@ void ZSetFamily::ZCount(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+vector<ScoredMap> OpDiff(EngineShard* shard, Transaction* t) {
+  ArgSlice keys = t->GetShardArgs(shard->shard_id());
+  DVLOG(1) << "shard:" << shard->shard_id() << ", keys " << vector(keys.begin(), keys.end());
+  DCHECK(!keys.empty());
+
+  vector<ScoredMap> results;
+  results.reserve(keys.size());
+
+  auto& db_slice = shard->db_slice();
+  for (size_t i = 0; i < keys.size(); ++i) {
+    auto it = db_slice.Find(t->GetDbContext(), keys[i], OBJ_ZSET);
+    if (!it) {
+      results.push_back({});
+      continue;
+    }
+
+    ScoredMap sm = FromObject((*it)->second, 1);
+    results.push_back(std::move(sm));
+  }
+
+  return results;
+}
+
+void ZSetFamily::ZDiff(CmdArgList args, ConnectionContext* cntx) {
+  vector<vector<ScoredMap>> maps(shard_set->size());
+  auto cb = [&](Transaction* t, EngineShard* shard) {
+    maps[shard->shard_id()] = OpDiff(shard, t);
+    return OpStatus::OK;
+  };
+
+  cntx->transaction->ScheduleSingleHop(std::move(cb));
+
+  const string_view key = ArgS(args, 1);
+  const ShardId sid = Shard(key, maps.size());
+  auto& sm = maps[sid];
+  if (sm.empty()) {
+    (*cntx)->SendEmptyArray();
+    return;
+  }
+  auto result = std::move(sm[0]);
+  sm.erase(sm.begin());
+  for (auto& vsm : maps) {
```

Comment:
We don't have a log in the first part: ScoredMap is a hashmap.

Where the log comes from, is keeping the sorted order with the scores. Redis stores szets as both hash + skiplist (alternative to BSTs), so they do `(N-K) * logN` for removals. We do a single `K log K` sort at the end.

So all in all this adds up to `L + K log K`

We can't easily do better, because we don't have a ready sorted order to delete from. I suggest not to over-engineer it for now as its really just a minor issue.

Redis impl for reference: 

https://github.com/redis/redis/blob/c2f1815bcbf5a02ae3920c2efa31099226a966bb/src/t_zset.c#L2443

Reply:
Right! sorry, I didn't notice you convert it to a hash map :)

Reply:
@dranikpg thank you for adding this before I did! 

also don't we do the same for the `zset` family ? https://github.com/dragonflydb/dragonfly/blob/main/src/redis/t_zset.c which I assume is the same file? (therefore same skiplist impl as redis)

Reply:
Yes, but I still suggest using FromObject and the ScoredMap for now. We don't have nice wrappers around the low-level structures (mostly the big visitor for access)